hail/src/main/scala/is/hail/expr/ir/InferPType.scala

```diff
@@ -140,204 +141,205 @@ object InferPType {
-        assert(compare.pType2.isOfType(PBoolean()))
+        assert(compare._pType2.isOfType(PBoolean()))
 
-        PArray(et, a.pType2.required)
+        PArray(et, a._pType2.required)
       }
       case ToSet(a) => {
         InferPType(a, env)
-        val et = coerce[PIterable](a.pType2).elementType
-        PSet(et, a.pType2.required)
+        val et = coerce[PIterable](a._pType2).elementType
+        PSet(et, a._pType2.required)
       }
       case ToDict(a) => {
         InferPType(a, env)
-        val elt = coerce[PBaseStruct](coerce[PIterable](a.pType2).elementType)
+        val elt = coerce[PBaseStruct](coerce[PIterable](a._pType2).elementType)
         // Dict key/value types don't depend on PIterable's requiredeness because we have an interface guarantee that
         // null PIterables are filtered out before dict construction
         val keyRequired = elt.types(0).required
         val valRequired =  elt.types(1).required
-        PDict(elt.types(0).setRequired(keyRequired), elt.types(1).setRequired(valRequired), a.pType2.required)
+        PDict(elt.types(0).setRequired(keyRequired), elt.types(1).setRequired(valRequired), a._pType2.required)
       }
       case ToArray(a) => {
         InferPType(a, env)
-        val elt = coerce[PIterable](a.pType2).elementType
-        PArray(elt, a.pType2.required)
+        val elt = coerce[PIterable](a._pType2).elementType
+        PArray(elt, a._pType2.required)
       }
       case ToStream(a) => {
         InferPType(a, env)
-        val elt = coerce[PIterable](a.pType2).elementType
-        PStream(elt, a.pType2.required)
+        val elt = coerce[PIterable](a._pType2).elementType
+        PStream(elt, a._pType2.required)
       }
       case GroupByKey(collection) => {
         InferPType(collection, env)
-        val elt = coerce[PBaseStruct](coerce[PStreamable](collection.pType2).elementType)
-        PDict(elt.types(0), PArray(elt.types(1)), collection.pType2.required)
+        val elt = coerce[PBaseStruct](coerce[PStreamable](collection._pType2).elementType)
+        PDict(elt.types(0), PArray(elt.types(1)), collection._pType2.required)
       }
       case ArrayMap(a, name, body) => {
         InferPType(a, env)
-        InferPType(body, env.bind(name, a.pType2.asInstanceOf[PArray].elementType))
-        coerce[PStreamable](a.pType2).copyStreamable(body.pType2, body.pType2.required)
+        InferPType(body, env.bind(name, a._pType2.asInstanceOf[PArray].elementType))
+        coerce[PStreamable](a._pType2).copyStreamable(body._pType2, body._pType2.required)
       }
       case ArrayZip(as, names, body, _) =>
-        as.foreach(InferPType(_, env))
-        InferPType(body, env.bindIterable(names.zip(as.map(_.pType2))))
-        coerce[PStreamable](as.head.pType2).copyStreamable(body.pType2, as.forall(_.pType2.required))
+        as.foreach(a => InferPType(a, env))
+
+        InferPType(body, env.bindIterable(names.zip(as.map(_._pType2.asInstanceOf[PArray].elementType))))
```

Comment:
This was a bug; ArrayZip expects environments in which element pTypes are bound, not the pTypes of the ArrayZip ir arguments themselves.